### ISSUE TITLE
Spaces support

### ISSUE THREAD

**Issue by j4mie** [jigish/slate#78]
First, Slate is fantastic. Exactly what I've been looking for.

One thing is keeping it from being perfect for me, though. My use case is that when I connect or disconnect an external monitor, I want to re-layout windows that are on multiple spaces. I have groups of windows that I always keep on particular spaces, and I want to be able to use layouts to reposition them automatically.

My current stab at a config file:

```
alias laptop 1366x768
alias 1-main  move screenOriginX;screenOriginY+5 screenSizeX-5;screenSizeY-10 ${laptop}
layout laptop 'Google Chrome':REPEAT ${1-main}
layout laptop 'Sublime Text 2':REPEAT ${1-main}
bind r:alt layout laptop
```

(I've just bound this to a keyboard shortcut for testing, I've not set up the defaults yet)

Chrome is on Desktop 1, and Sublime is on Desktop 2. If I hit ctrl+r while on Desktop 1, it correctly repositions Chrome, but doesn't touch Sublime. If I'm on Desktop 2, Sublime works but Chrome doesn't.

Is there any way to specify which Space to put windows on? Apologies if I'm missing something, but I can't figure out any way of doing this from the documentation.

Cheers!

**Comment by jigish** [comment]
Hey! Actually right now there isn't spaces support but I'm getting more and more requests to add it. I'll see what I can do. Look for a comment on this when I figure it out :)

**Comment by tomusher** [comment]
This feature would be fantastic - being able to configure all my spaces based on my current 'mode' (development, writing, emails, etc.) would be really useful. Even better would be able to trigger these layouts from the command line so I could use Alfred or any other tool to trigger them.

**Comment by iwinux** [comment]
I'm also interested in this feature, and here's some references that may be usefuly:
1. http://stackoverflow.com/questions/2308820/how-can-i-programatically-move-one-terminal-app-window-to-another-space
2. http://stackoverflow.com/questions/7049344/moving-a-window-to-a-new-space-in-mission-control

**Comment by justinjones** [comment]
Would also love this!

**Comment by jigish** [comment]
Huh. Looks like I'm shit out of luck: http://thecocoabots.com/blog/post/377/

The Apple banhammer strikes again!

I won't give up on this feature as I would really love it myself, but unfortunately I was thinking I'd try to throw together some applescript to do this as programmatically it doesn't seem possible with Apple's current APIs and according to the Hyperspaces dev team this is no longer possible.

I'll keep my eyes (and this ticket) open. Hopefully I'll stumble across something that works for Lion/Mountain Lion. If any of you come across something please let me know.

As of right now:

OS X 10.5 and 10.6 can technically support this through applescript.
OS X 10.7+ does not support this in any way.

As such, I don't feel comfortable implementing a feature that is only usable by the older version of OS X so yah, I'll be shelving this idea until I come across a solution (if ever i do) =/.

Sorry guys. Wish the answer wasn't such a fail.

**Comment by zlandau** [comment]
This may have horrible performance, but I wonder how hard it would be to implement 'virtual spaces'. So Slate would keep track of what is on each virtual space, and when you switch one it basically just minimizes the current applications and lays out the applications on the new slate? It might be too slow to be useful, but it might be worth a shot.

**Comment by jigish** [comment]
Hmm. good idea. I can try playing around with it. Its going to be lower priority though as I think it'll be a bit more trouble than it is worth.

**Comment by tomusher** [comment]
That's a pity, can only hope official APIs exist one day.

To avoid hacky solutions getting in to Slate, perhaps you could (through your plan for terminal support), give users some tools to script their own ways around it. For example, if it's possible to force a window to the current space, we could do something like:

```
trigger slate layout foo
press keys cmd+right
trigger slate layout bar # this layout forces windows on to the current space
```

**Comment by jigish** [comment]
Yah, thats the plan. Issue #85 will track the command line support. Ideally I'd just allow triggering bindings and such from command-line and users can create their own shell scripts that they can keybind. I can potentially even add the ability to keybind scripts directly in the .slate file (now that I think about it, that sounds like a great idea...).

**Comment by joaogeada** [comment]
There is a set key bindings that will cause a window whose title is clicked & held to be moved to another space: 
Ctrl-&lt;num&gt;
So Ctrl-1 while clicking & holding the title of a window will cause that window to move to the first Space.
If you could synthetically generate the same set of events from within Slate you'd effectively have the required behavior without needing a Spaces API per-se. Yeah, I know, working around arbitrary Apple limitations, but better than nothing :-)

**Comment by ansother** [comment]
sizeup seems to have figured out a way. It works on 10.8 now.

**Comment by romaimperator** [comment]
It looks like you can get the current space ID, (http://stackoverflow.com/questions/554380/how-to-detect-which-space-the-user-is-on-in-mac-os-x-leopard and a more recent discussion http://stackoverflow.com/questions/6768684/osx-lion-applescript-how-to-get-current-space-from-mission-control) which could be used in the configuration to identify which space you want windows on just like can currently be done with multiple monitors. Another app I use, TotalSpaces also can move windows between spaces so there must be a way.

A temporary solution could be to trigger a specific layout when the current space changes. I found an old and probably no longer useful post on that http://tonyarnold.com/post/2023/detecting-when-the-active-space-changes/. Along with this, the above space IDs could be used to limit which windows are changed depending on which space the user is on. Then you could configure an app such as Finder to be in one position and size on one space but another position or size on a different space.

I'd also like to put in my +1 on this feature. A friend of mine has been using Awesome WM on his ArchLinux box and tempting me to try and find something similar for OS X and your app is the closest I've found. The apps I've tried all suck because it takes too much time to configure where the apps are. Then if your screen changes resolution, the positions are all messed up and you have to redo all that work. At least after I spend time writing the config for Slate I just need to press a key combo to put everything in its place so thanks for writing this!

**Comment by wmayner** [comment]
This feature would be incredible - I've been desperately searching for something that can do this.

**Comment by brabalan** [comment]
osxmonad recently added space support. Their commit has some objective C files that shows how to access some space information.
https://github.com/xmonad/osxmonad/commit/66b1f7431d60287d4449dec2e2251a4ea3a96ff6

**Comment by sbc** [comment]
SizeUp has this feature, and it works fine on Lion. Specifically they offer:

> Spaces Actions:
> 
> > Space Prev   Move a window to the previous Space
> > Space Next   Move a window to the next Space
> > Space Above  Move a window to the Space above
> > Space Below  Move a window to the Space below 

Would be great to have the equivalent for Slate.

**Comment by mccarthyjames** [comment]
+1 for this functionality. I've been using both Divvy and SizeUp and Slate is a great alternative to them both, but I do use the space switching functioned mentioned above.

**Comment by ejoubaud** [comment]
+1, it would be a huge win fo this remarkable piece of software :)

**Comment by indrekj** [comment]
+1

**Comment by adrianpike** [comment]
+1 :)

**Comment by chanpory** [comment]
+1

**Comment by slinkp** [comment]
+1 to spaces support.  -1 to apple for removing useful APIs.

**Comment by ckruse** [comment]
+1

**Comment by bluntpeak** [comment]
+1

**Comment by markroland** [comment]
+1

**Comment by asellitt** [comment]
+1

**Comment by olegbl** [comment]
+1

**Comment by wamatt** [comment]
+1.

**Comment by wamatt** [comment]
FWIW, I went looking for WM for OSX with Spaces support. It seems an alpha version of Xmonad for OSX supports this. https://github.com/xmonad/osxmonad

Might be worth checking out their approach?

**Comment by ianyh** [comment]
`osxmonad` relies on private APIs, which it sounds like @jigish wants to avoid. I've been working on an xmonad-style tiling window manager for OS X (similar to `osxmonad`, but in Objective-C instead of Haskell, you can find it [here](https://github.com/ianyh/Amethyst) if you're curious). I figured out a way to move windows between spaces and have a writeup on how it works [here](http://ianyh.github.io/blog/2013/06/05/accessibility/), which might be of interest here.

**Comment by dpo** [comment]
+1

**Comment by wamatt** [comment]
Thanks for the explanation @ianyh

**Comment by ckruse** [comment]
Hey,

On 13/11/13 11:05, md143rbh7f wrote:

> Hi everyone, I'm working on implementing virtual spaces in .slate.js a la
> https://github.com/jigish/slate/issues/78#issuecomment-9120241
> I intend for it to function somewhat like in xmonad. I can post the code later if anyone is interested.

Please, don't. If you can't get slate working with the default OS X
spaces then talk to the TotalSpaces people, maybe they can provide an
API. But I don't think we need another pseudo-spaces implementation.

Best regards,
 Christian Kruse

**Comment by mrooney** [comment]
:+1: Yeah, this would be awesome to see; it is the only thing I need SizeUp for currently. It must be possible since SizeUp can move windows to the next and previous space in 10.5+ including 10.8/9. Even if it requires a private / undocumented API which could stop working at some future time, that seems better than not working at all presently :)

**Comment by sprig** [comment]
Hello,

@ckruse, I disagree with you for the following reasons:
0. Such a feature would be useful in general as you would not be reliant on the whims of Apple in regards to what UI they decide next is best for you (me).
1. TotalSpaces is neither free nor open source, with all this entails, so it would be great to have this functionality in an open source program.
2. Imho slate, being as far as it seems to me currently, the best "window manager"ish thing would be the natural place to put such functionality, traditionally found in window managers.

Cheers

**Comment by mattes** [comment]
:+1:  for spaces support

**Comment by d6e** [comment]
:+1: I'd be happy if I could just bind some keys to move a window from one space to the next.

**Comment by elads** [comment]
Definitely +1
Amazing work on Slate. I am late to the party but it is a pretty awesome tool so far.

The application "Stay" http://cordlessdog.com/stay/ does some of these things while working with Spaces. I set it up to see if it could handle saving window layouts that are activated by a change of Monitor connection. It did a pretty good job, and did work with Spaces. but unfortunately it was too buggy to deserve a 15$ buy.

My use case is a laptop which is connected and disconnected from a desk monitor of a different resolution many times throughout my work day and I come in and out of meetings. I set up Stay to optimize the location and placement of a handful of always open applications in their respective Space for each monitor's resolution. In addition to Slate's other handy features, I am hoping to use it to mimic the same function.  (I use TotalSpaces as an enhancement to the native Maverick's Spaces) 

Thanks for the great work on Slate. I look forward to see it evolve.

**Comment by franzwr** [comment]
People,

I've just discovered Slate and love it, but can't stand moving my windows around spaces manually! So I've created a workaround hack using the [Cliclick](https://github.com/BlueM/cliclick) tool, which can send keyboard and mouse input to OS X from the command line. You can install it using Homebrew: `brew install cliclick`. The point is to emulate clicking the title bar and pressing Cmd-arrow.

I defined the following actions in `.slate.js`:

```
function moveWindowToPreviousSpace(win) {
  var windowClickCoords = (win.rect().x + win.rect().width/2) + "," + (win.rect().y + 5);
  S.shell(
    "/usr/local/bin/cliclick -r c:" + windowClickCoords +
    " w:50 dd:" + windowClickCoords + 
    " kd:ctrl kp:arrow-left w:500 ku:ctrl du:" + windowClickCoords
  );
};

function moveWindowToNextSpace(win) {
  var windowClickCoords = (win.rect().x + win.rect().width/2) + "," + (win.rect().y + 5);
  S.shell(
    "/usr/local/bin/cliclick -r c:" + windowClickCoords +
    " w:50 dd:" + windowClickCoords + 
    " kd:ctrl kp:arrow-right w:500 ku:ctrl du:" + windowClickCoords
  );
};
```

Caveats:
- The window click coordinates and delays are found by trial and error, you may need to adjust them.
- The tool only allows for sending a limited range of keystrokes, so moving window to desktop N isn't possible with it.

Hope it helps!

**Comment by mattr-** [comment]
I started looking into adding Spaces support for Slate last night. Looks doable but I need some more time to figure it out.

**Comment by sdegutis** [comment]
The discussion here may help: sdegutis/hydra#116

**Comment by wamatt** [comment]
@sdegutis looks good but also hacky. I'm curious though, is using a private api that bad?

**Comment by ianyh** [comment]
It's a trade off. I originally came up with the non-private API method for moving windows between spaces because I wanted to be able to submit to the Mac AppStore. If you don't care about doing that then the only reason to not use private APIs is fragility. You run the risk of having the re-reverse engineer the private APIs all over again every time a new version of OS X is released.

**Comment by sdegutis** [comment]
@wamatt Good point. I like using private APIs personally, because they're generally more reliable within a single OS X version. In fact I think I'll look at what osxmonad does.

**Comment by sdegutis** [comment]
@ianyh I don't think the Mac App Store accepts apps that require accessibility anyway (I know Divvy was grandfathered in).

**Comment by ianyh** [comment]
@sdegutis Huh. That is an interesting data point, but I have actually already moved towards using private APIs in Amethyst and given up on being in the AppStore. There's just not enough public API support for window management.

**Comment by sdegutis** [comment]
@ianyh yep; a few days ago I started just using private APIs like crazy in Hydra. On the bright side, they actually work :D (at least in 10.9, we'll see about 10.10)

**Comment by sdegutis** [comment]
Further discussion with @puffnfresh suggests that the existing private APIs worked on 10.8 but broke on 10.9...

**Comment by wolfbiter** [comment]
+1 on wanting this feature

**Comment by yn5** [comment]
I'm already completely in love with Slate but this little thing would make it absolutely perfect! So +1 for me as well.

**Comment by enoch** [comment]
Used to use SizeUp, decided to use Slate instead when I picked up my new MBA and I'm loving it! The only thing tempting me back to SizeUp right now is the lack of Spaces/MissionControl support so +1! :+1:

**Comment by Konafets** [comment]
:+1:

**Comment by wamatt** [comment]
"Slate with spaces support" and used with Totalspaces would be a really sweet setup.

**Comment by KMontag42** [comment]
:+1:

**Comment by MoOx** [comment]
FYI I created an issue on a maintained fork https://github.com/mattr-/slate/issues/8
(See #437)

**Comment by xbeta** [comment]
+1 please

**Comment by stolendog** [comment]
+1

**Comment by alexilyaev** [comment]
@jigish Hey, it's been a while, what's the current status of this issue?

**Comment by sdegutis** [comment]
lolololololol

**Comment by alexilyaev** [comment]
@sdegutis Oh, ok, should have read #437 :-)
Adding the question to [mattr-#8](https://github.com/mattr-/slate/issues/8)

**Comment by kevinmtrowbridge** [comment]
Hello guys,

I am just like you -- over the years I have gotten used to having certain "window management keyboard shortcuts" and, each time I upgrade OSX essentially, they have screwed around with it and broken my existing solutions.  Literally every time I have upgraded, I've had to rejigger my setup.

Yesterday I upgraded from Mavericks -> El Capitan ... this broke "TotalSpaces 2" ... which I was using for spaces management, and for moving windows from one space to another with a keyboard shortcut.  I'm ok with El Capitan's new Spaces feature (although it is not as good as TotalSpaces) ... but amazingly in El Capitan there's no keyboard shortcut to move a window to a new space.

I dug and dug, and I discovered that in "SizeUp" (which I have used in the past and still have a license to), the keyboard shortcuts to move windows between spaces, are working.  

So -- really this doesn't have anything to do with Slate, but in my research to try and figure out how I can accomplish this _at all_ on El Capitan, I ended up here ... and so I am just leaving this here, to help folks upgrading to El Capitan in the future: try SizeUp.

http://www.irradiatedsoftware.com/sizeup/

Best,
Kevin
